lib/common/common/src/cpu.rs

```diff
@@ -1,7 +1,196 @@
+            return None;
+        }
+
+        // Try to acquire selected number of CPUs
+        let result = Semaphore::try_acquire_many_owned(self.semaphore.clone(), num_cpus);
+        let permit = match result {
+            Ok(permit) => permit,
+            Err(TryAcquireError::NoPermits) => return None,
+            Err(TryAcquireError::Closed) => unreachable!("Cannot acquire CPU permit because CPU budget semaphore is closed, this should never happen"),
+        };
+
+        Some(CpuPermit::new(num_cpus, permit))
+    }
+
+    /// Check if there is enough CPU budget available for the given `desired_cpus`.
+    ///
+    /// This checks for the minimum number of required permits based on the given desired CPUs,
+    /// based on `min_permits`. To check for an exact number, use `has_budget_exact` instead.
+    ///
+    /// A desired CPU count of `0` will always return `true`.
+    pub fn has_budget(&self, desired_cpus: usize) -> bool {
+        self.has_budget_exact(self.min_permits(desired_cpus))
+    }
+
+    /// Check if there are at least `budget` available CPUs in this budget.
+    ///
+    /// A budget of `0` will always return `true`.
+    fn has_budget_exact(&self, budget: usize) -> bool {
+        self.semaphore.available_permits() >= budget
+    }
+
+    /// Wait until we have CPU budget available for the given number of desired CPUs.
+    ///
+    /// Waits for at least the minimum number of permits based on the given desired CPUs. For
+    /// example, if `desired_cpus` is 8, this will wait for at least 4 to be available. See
+    /// [`Self::min_permits`].
+    ///
+    /// - `1` to wait for any CPU budget to be available.
+    /// - `0` will always return immediately.
+    ///
+    /// Uses an exponential backoff strategy up to 10 seconds to avoid busy waiting.
+    pub async fn wait_until_budget(&self, desired_cpus: usize) {
+        let min_required = self.min_permits(desired_cpus);
+        if self.has_budget_exact(min_required) {
+            return;
+        }
+
+        // Wait for CPU budget to be available with exponential backoff
+        // TODO: find better way, don't busy wait
+        let mut delay = Duration::from_micros(100);
+        while !self.has_budget_exact(min_required) {
+            time::sleep(delay).await;
+            delay = (delay * 2).min(Duration::from_secs(10));
+        }
+    }
+}
+
+impl Default for CpuBudget {
+    fn default() -> Self {
+        Self::new(get_cpu_budget(0))
+    }
+}
+
+/// CPU permit, used to limit number of concurrent CPU-intensive operations
+///
+/// This permit represents the number of CPUs allocated for an operation, so that the operation can
+/// respect other parallel workloads. When dropped or `release()`-ed, the CPUs are given back for
+/// other tasks to acquire.
+///
+/// These CPU permits are used to better balance and saturate resource utilization.
+pub struct CpuPermit {
+    /// Number of CPUs acquired in this permit.
+    pub num_cpus: u32,
```

Comment:
Maybe we can use a `u16` instead to represent the number of cpu :)

Reply:
have you seen those nvidia monsters? 
![image](https://github.com/qdrant/qdrant/assets/1935623/ca8097bb-c6c2-4c35-8ece-acb470843829)

Reply:
I used a `u32` because we pass this type to the semaphore when claiming permits: https://docs.rs/tokio/latest/tokio/sync/struct.Semaphore.html#method.acquire_many

Do you think that's a good enough reason?

Reply:
Let's keep `u32` :+1: